### PR TITLE
[Serializer] Normalizers can serialize collections and scalars

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| License | MIT |
| Doc PR | n/a |

Currently, the logic for serializing `array` and scalars is hardcoded in the serializer. This is not possible to have a custom serializer normalizing collections. This a big limitation, for instance it's not possible to create an normalizer creating [Hydra collections](http://www.hydra-cg.com/spec/latest/core/#h-collections) for a PHP array.

This PR fix that.

### CODE REVIEW

**Comment by jakzal** [comment]
Would you be able to cover this with tests?

**Comment by dunglas** [comment]
@jakzal tests added.

File: src/Symfony/Component/Serializer/Serializer.php [hunk 1]
```diff
@@ -179,14 +185,18 @@ public function supportsDenormalization($data, $type, $format = null)
      */
     private function getNormalizer($data, $format)
     {
-        $class = get_class($data);
-        if (isset($this->normalizerCache[$class][$format])) {
-            return $this->normalizerCache[$class][$format];
+        if ($isObject = is_object($data)) {
```
Comment (stof) [inline src/Symfony/Component/Serializer/Serializer.php]: if you can pass something else than an object to this method, you need to update the phpdoc

**Comment by stof** [comment]
For Traversable objects, it is already possible to customize it

**Comment by dunglas** [comment]
@stof right. Fixed the PR description and the PHPDoc.

**Comment by dunglas** [comment]
The Travis failure seems unrelated with this PR.

**Comment by fabpot** [comment]
Thank you @dunglas.

**Pull request closed without merging**